### DISCUSSION TITLE
How to set loading state or error state for SSR data in Next 12v ?

### DISCUSSION THREAD

**Discussion by thanhtutzaw** [vercel/next.js#46824] [Help]
### I was using SSR data fetching from Firebase . Normally it is fast and no need to add loading state . But when I don't open the app for really long time in Production . The data is not loaded , it need to wait may be 40s and 3 time reload . Why it is happening in SSR and how to handle in Next.js ? I also implemented authentication with next-firebase-auth . Is it related ?

**Comment by icyJoseph** [comment]
Could it be that your firebase instance or something like that, is shutdown/killed when inactive? And then it has to be reactivated by firebase? It'd apply to authentication too.

When you say SSR, you mean GetServerSideProps? You could run a Promise.race, your data fetching vs a timeout, if your timeout runs out, you can return props to the UI that make it load a spinner. 

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race

You'd have to cancel the on-going async work with an AbortController, and let the UI render the spinner. Though cancelling the request maybe sends firebase to sleep again?

In the page with spinner, you could poll for data... until firebase is awake for example, if that's the case.

Might post some pseudo-code later :) but let me know what you think.

**Reply by thanhtutzaw** [reply]
How can I run promise.race in ssr . Here is some code

`export const getServerSideProps = withAuthUserTokenSSR()( async ({AuthUser}) => {

let notes = null;
const id = AuthUser.id;
const q = collection(db,users/${id}/notes)
const docSnap =await getDocs(q)

notes = docSnap.docs.map(doc => ({
id:doc.id,
...doc.data()
}))

return{
props:{
notes
}
}
})`
